sierra2mlir/src/types/mod.rs

```diff
@@ -93,6 +93,76 @@ impl<'ctx> Compiler<'ctx> {
                         SierraType::Struct { ty: struct_type, field_types: types },
                     );
                 }
+                "Enum" => {
+                    // https://mapping-high-level-constructs-to-llvm-ir.readthedocs.io/en/latest/basic-constructs/unions.html#tagged-unions
+
+                    // in case the enum has variants with the same type, just store a uniqued one.
+                    let mut unique_enum_types = HashSet::new();
+                    // all the enum variant types, even repeated, for later use when storing the variant value.
+                    let mut enum_variant_types = vec![];
+
+                    let _user_type = match &type_decl.long_id.generic_args[0] {
+                        GenericArg::UserType(x) => x,
+                        _ => {
+                            unreachable!("first arg on struct libfunc should always be a user type")
+                        }
+                    };
+
+                    for gen_arg in type_decl.long_id.generic_args.iter().skip(1) {
+                        let gen_arg_type_id = match gen_arg {
+                            GenericArg::Type(gen_arg_typeid) => gen_arg_typeid.id.to_string(),
+                            _ => unreachable!("should always be a type"),
+                        };
+                        unique_enum_types.insert(gen_arg_type_id.clone());
+                        enum_variant_types.push(
+                            storage
+                                .types
+                                .get(&gen_arg_type_id)
+                                .cloned()
+                                .expect("type should exist"),
+                        );
+                    }
+
+                    let enum_variant_sierra_types = unique_enum_types
+                        .iter()
+                        .map(|x| storage.types.get(x).cloned().expect("type should exist"))
+                        .collect_vec();
+                    let unique_enum_variant_types =
+                        enum_variant_sierra_types.iter().map(SierraType::get_type).collect_vec();
+
+                    // The enum is a struct consiting of the tag and enough memory to allocate the biggest element.
```

Comment:
Maybe it'd be nice to mention C unions, as it's basically the same.

Reply:
i think the link in a comment already mentions it